idx/search/types.go

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package search
+
+import (
+	"github.com/m3db/m3ninx/idx/doc"
+	"github.com/m3db/m3ninx/idx/index"
+)
+
+// Query is a search query for documents.
+type Query interface {
```

Comment:
Thought about our chat on this some more. Feels like it'd be a better abstraction to have a Query composed of some other type (e.g. Filter/Predicate). And that type has the `Execute` method. That way the searcher is responsible for the composition of results, where as the sub-type can be delegated the responsibility of executing a type of query. 

What do you think?

Reply:
I'm not sure I completely understand what you mean here. Queries can be composed of other queries, for example the `conjuctionQuery` is composed of a slice of queries and is responsible for executing them and taking their intersection so I'm not sure I see what the benefit of another type would be.

Reply:
The difference in the two approaches being without `Filter`(or w/e), `Query.Execute` has to execute all its sub-components (or atleast that's what the signature would imply). The whole point of the `Searcher` interface is to do that, no?